embassy-nrf/src/saadc.rs

```diff
@@ -125,29 +119,33 @@ impl<'d, T: PositivePin> OneShot<'d, T> {
     }
 }
 
-impl<'d, T: PositivePin> Drop for OneShot<'d, T> {
+impl<'d> Drop for OneShot<'d> {
     fn drop(&mut self) {
         let r = self.regs();
         r.enable.write(|w| w.enable().disabled());
     }
 }
 
 pub trait Sample {
-    type SampleFuture<'a>: Future<Output = i16> + 'a
+    type SampleFuture<'a, T>: Future<Output = i16> + 'a
     where
+        T: PositivePin + 'a,
         Self: 'a;
 
-    fn sample<'a>(&'a mut self) -> Self::SampleFuture<'a>;
+    fn sample<'a, T: PositivePin>(&'a mut self, pin: &'a T) -> Self::SampleFuture<'a, T>;
 }
 
-impl<'d, T: PositivePin> Sample for OneShot<'d, T> {
+impl<'d> Sample for OneShot<'d> {
     #[rustfmt::skip]
-    type SampleFuture<'a> where Self: 'a = impl Future<Output = i16> + 'a;
+    type SampleFuture<'a, T> where Self: 'a, T: PositivePin + 'a = impl Future<Output = i16> + 'a;
```

Comment:
You're using the "full" GATs here (generic over a type, not over a lifetime). We should avoid using these, since it's very likely that lifetime-GATs will be stabilized soon and full-GATs will take a long while more.

Maybe it's workaroundable by calling `pin.channel()` outside the `async move{}` block and moving the integer value inside, so that the future type doesn't depend on T.